# Table of Content:
* [Li Yuan's second week work](#2)
* [Li Yuan's third week work](#3)
    * [Pandas convert](#31)
    * [MdTraj convert](#32)

<a id='2'></a>
# Li Yuan's second week work 

This is a set of basic examples of the usage and outputs of the various individual functions included in. There are generally three types of functions:

+ Functions that perform searches and return lists of PDB IDs
+ Functions that get information about specific PDB IDs
+ Other general-purpose lookup functions

The list of supported search types, as well as the different types of information that can be returned for a given PDB ID, is large (and growing) and is enumerated in the docstrings of pypdb.py. The PDB allows a very wide range of different types of queries, and so any option that is not currently available can likely be implemented based on the structure of the query types that have already been implemented. Please submit feedback and pull requests on GitHub.

### I didn't find any funcion in that package pypdb we can use to extract seqres and atom, so I only use get_pdb_file() function from that package to get the file and write my own function to do that.

### Preamble

We import this package pypdb and prepare some other things.

In [3]:
%pylab inline
from IPython.display import HTML

## Import from local directory
import sys
sys.path.insert(0, '../pypdb')
from pypdb import *

## Import from installed package
# from pypdb import *

import pprint

%load_ext autoreload
%autoreload 2

Populating the interactive namespace from numpy and matplotlib


## This function I wrote is to extract only the seqres as a list

In [4]:
def get_seqres(pdb_id):
    """ Return the seqres sequence of a pdb file
    
    >>> get_seqres('4Z0L')
    >>> get_seqres('4lza')
    """
    pdb_file = get_pdb_file(pdb_id, filetype='pdb', compression=False) 
    # using a get_pdb_file() function from pypdb package to return a file with format 'pdb'.
    file1 = pdb_file.splitlines()
    # split this long string into list by \n.
    list_se = []
    for line in file1:
        if line[:6] == "SEQRES":
            list_se.append(line)
    return(list_se)

In [5]:
get_seqres('4lza')[:20]

['SEQRES   1 A  195  MSE HIS HIS HIS HIS HIS HIS SER SER GLY VAL ASP LEU          ',
 'SEQRES   2 A  195  GLY THR GLU ASN LEU TYR PHE GLN SER MSE THR LEU GLU          ',
 'SEQRES   3 A  195  GLU ILE LYS MSE MSE ILE ARG GLU ILE PRO ASP PHE PRO          ',
 'SEQRES   4 A  195  LYS LYS GLY ILE LYS PHE LYS ASP ILE THR PRO VAL LEU          ',
 'SEQRES   5 A  195  LYS ASP ALA LYS ALA PHE ASN TYR SER ILE GLU MSE LEU          ',
 'SEQRES   6 A  195  ALA LYS ALA LEU GLU GLY ARG LYS PHE ASP LEU ILE ALA          ',
 'SEQRES   7 A  195  ALA PRO GLU ALA ARG GLY PHE LEU PHE GLY ALA PRO LEU          ',
 'SEQRES   8 A  195  ALA TYR ARG LEU GLY VAL GLY PHE VAL PRO VAL ARG LYS          ',
 'SEQRES   9 A  195  PRO GLY LYS LEU PRO ALA GLU THR LEU SER TYR GLU TYR          ',
 'SEQRES  10 A  195  GLU LEU GLU TYR GLY THR ASP SER LEU GLU ILE HIS LYS          ',
 'SEQRES  11 A  195  ASP ALA VAL LEU GLU GLY GLN ARG VAL VAL ILE VAL ASP          ',
 'SEQRES  12 A  195  ASP LEU LEU ALA THR GLY GLY THR ILE TYR ALA 

In [6]:
get_seqres('4Z0L')[:20]

['SEQRES   1 A  587  ALA ASN PRO CYS CYS SER ASN PRO CYS GLN ASN ARG GLY          ',
 'SEQRES   2 A  587  GLU CYS MET SER THR GLY PHE ASP GLN TYR LYS CYS ASP          ',
 'SEQRES   3 A  587  CYS THR ARG THR GLY PHE TYR GLY GLU ASN CYS THR THR          ',
 'SEQRES   4 A  587  PRO GLU PHE LEU THR ARG ILE LYS LEU LEU LEU LYS PRO          ',
 'SEQRES   5 A  587  THR PRO ASN THR VAL HIS TYR ILE LEU THR HIS PHE LYS          ',
 'SEQRES   6 A  587  GLY VAL TRP ASN ILE VAL ASN ASN ILE PRO PHE LEU ARG          ',
 'SEQRES   7 A  587  SER LEU ILE MET LYS TYR VAL LEU THR SER ARG SER TYR          ',
 'SEQRES   8 A  587  LEU ILE ASP SER PRO PRO THR TYR ASN VAL HIS TYR GLY          ',
 'SEQRES   9 A  587  TYR LYS SER TRP GLU ALA PHE SER ASN LEU SER TYR TYR          ',
 'SEQRES  10 A  587  THR ARG ALA LEU PRO PRO VAL ALA ASP ASP CYS PRO THR          ',
 'SEQRES  11 A  587  PRO MET GLY VAL LYS GLY ASN LYS GLU LEU PRO ASP SER          ',
 'SEQRES  12 A  587  LYS GLU VAL LEU GLU LYS VAL LEU LEU ARG ARG 

### This function I wrote is to extract only the atom sequence as a list

In [7]:
def get_atom(pdb_id):
    """ Return the atom sequence of a pdb file
    
    >>> get_atom('4Z0L')
    >>> get_atom('4lza')
    """
    pdb_file = get_pdb_file(pdb_id, filetype='pdb', compression=False)
    # using a get_pdb_file() function from pypdb package to return a file with format 'pdb'.
    file1 = pdb_file.splitlines()
    list_atom = []
    for line in file1:
        if line[:4] == "ATOM":
            list_atom.append(line)
    return(list_atom)

In [8]:
get_atom('4Z0L')[:10]

['ATOM      1  N   ALA A  33     113.744  17.524  85.910  1.00 75.99           N  ',
 'ATOM      2  CA  ALA A  33     114.749  17.116  86.884  1.00 76.70           C  ',
 'ATOM      3  C   ALA A  33     115.677  18.275  87.231  1.00 73.52           C  ',
 'ATOM      4  O   ALA A  33     116.176  18.367  88.354  1.00 75.48           O  ',
 'ATOM      5  CB  ALA A  33     115.548  15.934  86.358  1.00 78.19           C  ',
 'ATOM      6  N   ASN A  34     115.906  19.154  86.261  1.00 67.98           N  ',
 'ATOM      7  CA  ASN A  34     116.747  20.327  86.469  1.00 63.43           C  ',
 'ATOM      8  C   ASN A  34     116.113  21.264  87.492  1.00 60.58           C  ',
 'ATOM      9  O   ASN A  34     115.006  21.756  87.287  1.00 61.30           O  ',
 'ATOM     10  CB  ASN A  34     116.983  21.058  85.144  1.00 63.09           C  ']

In [9]:
get_atom('4lza')[:10]

['ATOM      1  N   THR A   0     -27.785   5.217 -21.426  1.00 50.53           N  ',
 'ATOM      2  CA  THR A   0     -27.459   5.049 -19.974  1.00 49.41           C  ',
 'ATOM      3  C   THR A   0     -25.949   5.130 -19.667  1.00 46.13           C  ',
 'ATOM      4  O   THR A   0     -25.572   5.789 -18.699  1.00 44.22           O  ',
 'ATOM      5  CB  THR A   0     -28.153   3.815 -19.346  1.00 51.85           C  ',
 'ATOM      6  OG1 THR A   0     -27.919   3.787 -17.932  1.00 52.21           O  ',
 'ATOM      7  CG2 THR A   0     -27.688   2.516 -19.989  1.00 53.52           C  ',
 'ATOM      8  N   LEU A   1     -25.087   4.511 -20.480  1.00 43.20           N  ',
 'ATOM      9  CA  LEU A   1     -23.681   4.942 -20.481  1.00 42.39           C  ',
 'ATOM     10  C   LEU A   1     -23.615   6.356 -21.059  1.00 43.21           C  ']

<a id='3'></a>
# Li Yuan's third week work

<a id='31'></a>
## We first used pandas to convert a list into dataframe 

### First we used split() to split each string in the list returned by get_atom() function 

In [10]:
import pandas as pd

In [11]:
def get_atom(pdb_id):
    """ Return the atom sequence of a pdb file as a pandas dataframe
    
    >>> get_atom('4Z0L')
    >>> get_atom('4lza')
    """
    pdb_file = get_pdb_file(pdb_id, filetype='pdb', compression=False)
    # using a get_pdb_file() function from pypdb package to return a file with format 'pdb'.
    file1 = pdb_file.splitlines()
    list_atom = []
    for line in file1:
        if line[:4] == "ATOM":
            list_atom.append(line)
    list_s_atom = [s.split() for s in list_atom]
    # split each string in a list by white spaces
    df = pd.DataFrame(list_s_atom)
    # use DataFrame function to convert a list to dataframe
    df["id"] = pdb_id
    # add one id column to exsiting dataframe
    return(df)

In [12]:
get_atom("4lza").head(11)

0   1    2    3  4  5        6      7        8     9     10 11    id
0   ATOM   1    N  THR  A  0  -27.785  5.217  -21.426  1.00  50.53  N  4lza
1   ATOM   2   CA  THR  A  0  -27.459  5.049  -19.974  1.00  49.41  C  4lza
2   ATOM   3    C  THR  A  0  -25.949  5.130  -19.667  1.00  46.13  C  4lza
3   ATOM   4    O  THR  A  0  -25.572  5.789  -18.699  1.00  44.22  O  4lza
4   ATOM   5   CB  THR  A  0  -28.153  3.815  -19.346  1.00  51.85  C  4lza
5   ATOM   6  OG1  THR  A  0  -27.919  3.787  -17.932  1.00  52.21  O  4lza
6   ATOM   7  CG2  THR  A  0  -27.688  2.516  -19.989  1.00  53.52  C  4lza
7   ATOM   8    N  LEU  A  1  -25.087  4.511  -20.480  1.00  43.20  N  4lza
8   ATOM   9   CA  LEU  A  1  -23.681  4.942  -20.481  1.00  42.39  C  4lza
9   ATOM  10    C  LEU  A  1  -23.615  6.356  -21.059  1.00  43.21  C  4lza
10  ATOM  11    O  LEU  A  1  -22.738  7.137  -20.688  1.00  41.35  O  4lza

## Second we want to put a couple of pdb entries into one dataframe.

In [32]:
def get_some_atom(L):
    """ Take a list with returning some atom parts of pdb files into one dataframe 
    
    >>> get_some_atom(["4lza", "4Z0L"])
    """
    frames = [get_atom(l) for l in L]
    return(pd.concat(frames))

## We test this function with a list ["4lza", "4Z0L]

In [31]:
get_some_atom(["4lza", "4Z0L"])

0      1    2    3  4    5        6       7        8     9     10  \
0      ATOM      1    N  THR  A    0  -27.785   5.217  -21.426  1.00  50.53   
1      ATOM      2   CA  THR  A    0  -27.459   5.049  -19.974  1.00  49.41   
2      ATOM      3    C  THR  A    0  -25.949   5.130  -19.667  1.00  46.13   
3      ATOM      4    O  THR  A    0  -25.572   5.789  -18.699  1.00  44.22   
4      ATOM      5   CB  THR  A    0  -28.153   3.815  -19.346  1.00  51.85   
5      ATOM      6  OG1  THR  A    0  -27.919   3.787  -17.932  1.00  52.21   
6      ATOM      7  CG2  THR  A    0  -27.688   2.516  -19.989  1.00  53.52   
7      ATOM      8    N  LEU  A    1  -25.087   4.511  -20.480  1.00  43.20   
8      ATOM      9   CA  LEU  A    1  -23.681   4.942  -20.481  1.00  42.39   
9      ATOM     10    C  LEU  A    1  -23.615   6.356  -21.059  1.00  43.21   
10     ATOM     11    O  LEU  A    1  -22.738   7.137  -20.688  1.00  41.35   
11     ATOM     12   CB  LEU  A    1  -22.743   4.008  -21.257  1.00  41.94   
12     ATOM     13   CG  LEU  A    1  -21.985   2.915  -20.487  1.00  42.50   
13     ATOM     14  CD1  LEU  A    1  -21.338   1.957  -21.475  1.00  44.31   
14     ATOM     15  CD2  LEU  A    1  -20.939   3.448  -19.508  1.00  39.42   
15     ATOM     16    N  GLU  A    2  -24.561   6.684  -21.950  1.00  45.78   
16     ATOM     17   CA  GLU  A    2  -24.686   8.048  -22.484  1.00  47.90   
17     ATOM     18    C  GLU  A    2  -24.917   9.038  -21.356  1.00  43.00   
18     ATOM     19    O  GLU  A    2  -24.419  10.162  -21.404  1.00  42.03   
19     ATOM     20   CB  GLU  A    2  -25.804   8.178  -23.531  1.00  53.41   
20     ATOM     21   CG  GLU  A    2  -25.853   9.566  -24.175  1.00  60.62   
21     ATOM     22   CD  GLU  A    2  -26.624   9.630  -25.488  1.00  72.10   
22     ATOM     23  OE1  GLU  A    2  -27.176   8.594  -25.927  1.00  73.33   
23     ATOM     24  OE2  GLU  A    2  -26.676  10.734  -26.086  1.00  76.05   
24     ATOM     25    N  GLU  A    3  -25.666   8.603  -20.347  1.00  41.47   
25     ATOM     26   CA  GLU  A    3  -25.920   9.409  -19.167  1.00  43.73   
26     ATOM     27    C  GLU  A    3  -24.644   9.647  -18.363  1.00  40.11   
27     ATOM     28    O  GLU  A    3  -24.351  10.791  -18.014  1.00  39.46   
28     ATOM     29   CB  GLU  A    3  -26.999   8.769  -18.294  1.00  51.67   
29     ATOM     30   CG  GLU  A    3  -28.420   9.010  -18.782  1.00  59.54   
...     ...    ...  ...  ... ..  ...      ...     ...      ...   ...    ...   
17862  ATOM  17866    C  SER  D  579  117.038  37.003   23.117  1.00  43.44   
17863  ATOM  17867    O  SER  D  579  116.636  37.790   22.260  1.00  42.16   
17864  ATOM  17868   CB  SER  D  579  116.247  36.940   25.494  1.00  51.78   
17865  ATOM  17869   OG  SER  D  579  116.200  35.525   25.465  1.00  52.65   
17866  ATOM  17870    N  PHE  D  580  117.259  35.718   22.871  1.00  40.53   
17867  ATOM  17871   CA  PHE  D  580  116.947  35.138   21.573  1.00  38.81   
17868  ATOM  17872    C  PHE  D  580  115.529  34.573   21.571  1.00  43.20   
17869  ATOM  17873    O  PHE  D  580  115.046  34.084   20.549  1.00  42.35   
17870  ATOM  17874   CB  PHE  D  580  117.966  34.058   21.207  1.00  35.80   
17871  ATOM  17875   CG  PHE  D  580  119.316  34.604   20.828  1.00  36.39   
17872  ATOM  17876  CD1  PHE  D  580  119.425  35.824   20.182  1.00  33.90   
17873  ATOM  17877  CD2  PHE  D  580  120.474  33.903   21.122  1.00  36.79   
17874  ATOM  17878  CE1  PHE  D  580  120.664  36.334   19.831  1.00  31.61   
17875  ATOM  17879  CE2  PHE  D  580  121.716  34.409   20.775  1.00  33.85   
17876  ATOM  17880   CZ  PHE  D  580  121.810  35.625   20.129  1.00  31.90   
17877  ATOM  17881    N  ASN  D  581  114.863  34.651   22.720  1.00  46.01   
17878  ATOM  17882   CA  ASN  D  581  113.467  34.239   22.817  1.00  48.70   
17879  ATOM  17883    C  ASN  D  581  112.585  35.332   23.416  1.00  46.12   
17880  ATOM  17884    O  ASN  D  581  113.068  36.218   24.1

## Next we want to put all ids into one dataframe

### We used get_all() to list all pdb entries. 

In [15]:
len(get_all())

169681

### We found there was 169681 entries in the current PDB  DataBase.

### We used concat() function to merge all the dataframe of each pdb entry into one huge dataframe.

In [16]:
frames = [get_atom(id) for id in get_all()[:2]]
pd.concat(frames)

0    1    2   3  4   5       6       7       8     9     10 11    id
0    ATOM    1  O5'   C  A   1  -4.549   5.095   4.262  1.00  28.71  O  100D
1    ATOM    2  C5'   C  A   1  -4.176   6.323   3.646  1.00  27.35  C  100D
2    ATOM    3  C4'   C  A   1  -3.853   7.410   4.672  1.00  24.41  C  100D
3    ATOM    4  O4'   C  A   1  -4.992   7.650   5.512  1.00  22.53  O  100D
4    ATOM    5  C3'   C  A   1  -2.713   7.010   5.605  1.00  23.56  C  100D
5    ATOM    6  O3'   C  A   1  -1.379   7.127   5.060  1.00  21.02  O  100D
6    ATOM    7  C2'   C  A   1  -2.950   7.949   6.756  1.00  23.73  C  100D
7    ATOM    8  O2'   C  A   1  -2.407   9.267   6.554  1.00  23.93  O  100D
8    ATOM    9  C1'   C  A   1  -4.489   7.917   6.825  1.00  20.60  C  100D
9    ATOM   10   N1   C  A   1  -4.931   6.902   7.826  1.00  19.25  N  100D
10   ATOM   11   C2   C  A   1  -4.838   7.263   9.158  1.00  16.72  C  100D
11   ATOM   12   O2   C  A   1  -4.287   8.308   9.505  1.00  15.49  O  100D
12   ATOM   13   N3   C  A   1  -5.367   6.448  10.085  1.00  15.96  N  100D
13   ATOM   14   C4   C  A   1  -5.978   5.310   9.736  1.00  16.84  C  100D
14   ATOM   15   N4   C  A   1  -6.592   4.588  10.676  1.00  19.14  N  100D
15   ATOM   16   C5   C  A   1  -6.059   4.907   8.376  1.00  17.68  C  100D
16   ATOM   17   C6   C  A   1  -5.522   5.732   7.461  1.00  17.68  C  100D
17   ATOM   18    P  DC  A   2  -0.178   6.220   5.647  1.00  24.85  P  100D
18   ATOM   19  OP1  DC  A   2   0.915   6.451   4.671  1.00  25.96  O  100D
19   ATOM   20  OP2  DC  A   2  -0.948   4.954   5.664  1.00  24.57  O  100D
20   ATOM   21  O5'  DC  A   2   0.435   6.502   7.097  1.00  24.10  O  100D
21   ATOM   22  C5'  DC  A   2   1.020   7.793   7.281  1.00  19.66  C  100D
22   ATOM   23  C4'  DC  A   2   1.034   8.184   8.738  1.00  17.99  C  100D
23   ATOM   24  O4'  DC  A   2  -0.290   8.244   9.222  1.00  17.23  O  100D
24   ATOM   25  C3'  DC  A   2   1.724   7.167   9.617  1.00  18.98  C  100D
25   ATOM   26  O3'  DC  A   2   3.130   7.395   9.564  1.00  18.39  O  100D
26   ATOM   27  C2'  DC  A   2   1.152   7.607  10.934  1.00  17.33  C  100D
27   ATOM   28  C1'  DC  A   2  -0.273   7.853  10.599  1.00  15.44  C  100D
28   ATOM   29   N1  DC  A   2  -1.070   6.635  10.823  1.00  14.48  N  100D
29   ATOM   30   C2  DC  A   2  -1.417   6.355  12.130  1.00  13.03  C  100D
..    ...  ...  ...  .. ..  ..     ...     ...     ...   ...    ... ..   ...
418  ATOM  460   N1  DC  B  23  14.637  21.541  22.637  1.00   5.21  N  101D
419  ATOM  461   C2  DC  B  23  15.618  22.342  22.078  1.00   6.76  C  101D
420  ATOM  462   O2  DC  B  23  16.700  21.866  21.792  1.00   6.72  O  101D
421  ATOM  463   N3  DC  B  23  15.317  23.677  21.932  1.00  10.02  N  101D
422  ATOM  464   C4  DC  B  23  14.115  24.232  22.250  1.00   7.10  C  101D
423  ATOM  465   N4  DC  B  23  13.904  25.543  22.033  1.00   0.00  N  101D
424  ATOM  466   C5  DC  B  23  13.108  23.372  22.777  1.00   7.82  C  101D
425  ATOM  467   C6  DC  B  23  13.414  22.064  22.930  1.00   4.87  C  101D
426  ATOM  468    P  DG  B  24  15.967  16.347  24.405  1.00  28.28  P  101D
427  ATOM  469  OP1  DG  B  24  16.200  14.964  23.832  1.00  27.79  O  101D
428  ATOM  470  OP2  DG  B  24  15.043  16.312  25.587  1.00  22.79  O  101D
429  ATOM  471  O5'  DG  B  24  17.290  17.159  24.722  1.00  28.58  O  101D
430  ATOM  472  C5'  DG  B  24  18.246  17.920  23.998  1.00  16.71  C  101D
431  ATOM  473  C4'  DG  B  24  18.914  18.907  24.951  1.00  12.60  C  101D
432  ATOM  474  O4'  DG  B  24  18.372  20.147  24.710  1.00   5.80  O  101D
433  ATOM  475  C3'  DG  B  24  18.654  18.701  26.432  1.00  14.53  C  101D
434  ATOM  476  O3'  DG  B  24  19.678  17.873  27.024  1.00  22.51  O  101D
435  ATOM  477  C2'  DG  B  24  18.593  20.091  27.036  1.00  12.59  C  101D
436  ATOM  478  C1'  DG  B  24  18.681  20.995  25.860  1.00   8.49  C  101D
437  ATOM  479   N9  DG  B  24  17.627  22.013  25.822  1.00  10.33  N  101D
438 

<a id='32'></a>
## We used mdtraj package to load pdb file into memory from URL

>MDTraj is a python library that allows users to manipulate molecular dynamics (MD) trajectories. Features include:
1. Wide MD format support, including pdb, xtc, trr, dcd, binpos, netcdf, mdcrd, prmtop, and more.
2. Extremely fast RMSD calculations (4x the speed of the original Theobald QCP).
3. Extensive analysis functions including those that compute bonds, angles, dihedrals, hydrogen bonds, secondary structure, and NMR observables.
4. Lightweight, Pythonic API.

In [17]:
import mdtraj as md # import this package

In [18]:
pdb = md.load_pdb("https://files.rcsb.org/view/4LZA.pdb")  # load data

In [19]:
print(pdb) # print to see how many frames and atoms, residues this file has 

<mdtraj.Trajectory with 1 frames, 2833 atoms, 512 residues, and unitcells>


## We convert this pdb file into topology

In [20]:
topology = pdb.topology

In [21]:
table, bonds = topology.to_dataframe()

In [22]:
print(table.head(7))

   serial name element  resSeq resName  chainID segmentID
0       1    N       N       0     THR        0          
1       2   CA       C       0     THR        0          
2       3    C       C       0     THR        0          
3       4    O       O       0     THR        0          
4       5   CB       C       0     THR        0          
5       6  OG1       O       0     THR        0          
6       7  CG2       C       0     THR        0          


In [23]:
topology.atom(10)

LEU1-O

In [24]:
topology.atoms

<generator object Topology.atoms at 0x7fe9cd22c390>

In [25]:
[i for i in topology.atoms][:10]

[THR0-N,
 THR0-CA,
 THR0-C,
 THR0-O,
 THR0-CB,
 THR0-OG1,
 THR0-CG2,
 LEU1-N,
 LEU1-CA,
 LEU1-C]

In [26]:
print(table.head(10))

   serial name element  resSeq resName  chainID segmentID
0       1    N       N       0     THR        0          
1       2   CA       C       0     THR        0          
2       3    C       C       0     THR        0          
3       4    O       O       0     THR        0          
4       5   CB       C       0     THR        0          
5       6  OG1       O       0     THR        0          
6       7  CG2       C       0     THR        0          
7       8    N       N       1     LEU        0          
8       9   CA       C       1     LEU        0          
9      10    C       C       1     LEU        0          


In [27]:
atom = pdb.atom_slice(range(2833))

In [28]:
print(atom)

<mdtraj.Trajectory with 1 frames, 2833 atoms, 512 residues, and unitcells>


In [29]:
atom.xyz

array([[[-2.7785,  0.5217, -2.1426],
        [-2.7459,  0.5049, -1.9974],
        [-2.5949,  0.513 , -1.9667],
        ...,
        [-0.6332, -1.3026, -0.3481],
        [-0.8265, -1.4563, -0.0902],
        [-2.8824,  1.244 , -0.1084]]], dtype=float32)

In [30]:
[i for i in topology.bonds][:10]

[Bond(THR0-CA, THR0-C),
 Bond(THR0-C, THR0-O),
 Bond(THR0-CA, THR0-CB),
 Bond(THR0-N, THR0-CA),
 Bond(THR0-CB, THR0-CG2),
 Bond(THR0-CB, THR0-OG1),
 Bond(THR0-C, LEU1-N),
 Bond(LEU1-CA, LEU1-C),
 Bond(LEU1-C, LEU1-O),
 Bond(LEU1-CA, LEU1-CB)]